File: examples-copy/docs/event-driven/events/event-system-ts/task-service.ts
import { IEventBus } from "./event-system-0";
import { NextSubtaskTriggeredEvent, StartSubtaskCommand, Task } from "./types";
import { TaskRepository } from "./repositories-0";
import { WorkspaceFileManager } from "./workspace-loader";

// Utility types
// type TaskUpdate = Partial<Task>;
// type TaskCreateParams = Omit<Task, "id" | "createdAt" | "updatedAt">;

class TaskService {
  constructor(
    private eventBus: IEventBus,
    private taskRepository: TaskRepository,
    private workspaceFileManager: WorkspaceFileManager
  ) {
    // Command handlers
    this.eventBus.subscribe<CreateTaskCommand>(
      "CreateTaskCommand",
      this.handleCreateTaskCommand.bind(this)
    );
    this.eventBus.subscribe<StartTaskCommand>(
      "StartTaskCommand",
      this.handleStartTaskCommand.bind(this)
    );

    // Event handlers
    // this.eventBus.subscribe<TaskCreatedEvent>(
    //   "TaskCreated",
    //   this.handleTaskCreated.bind(this)
    // );
    // this.eventBus.subscribe<TaskFolderCreatedEvent>(
    //   "TaskFolderCreated",
    //   this.handleTaskFolderCreated.bind(this)
    // );
  }

  private async handleCreateTaskCommand(
    command: CreateTaskCommand
  ): Promise<void> {
    const taskId = crypto.randomUUID();

    const task: Task = {
      id: taskId,
      seqNumber: 0,
      title: "This is a task",
      status: "CREATED",
      folderPath: "",
      subtasks: [
        {
          id: crypto.randomUUID(),
          taskId: taskId,
          team: {
            agent: "@assistant",
            human: null,
          },
          stepNumber: 0,
          title: "Planning",
          description: "Initial planning phase",
          status: "pending",
          inputType: "string",
          outputType: "json",
        },
        {
          id: crypto.randomUUID(),
          taskId: taskId,
          team: {
            agent: "@function_executor",
            human: "@user",
          },
          stepNumber: 1,
          title: "Setup",
          description: "Setup initial configuration",
          status: "pending",
          inputType: "json",
          outputType: "json",
        },
      ],
      config: command.taskConfig,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    };

    // Create task folder
    const folderPath = await this.workspaceFileManager.createTaskFolder(task);
    task.folderPath = folderPath;

    await this.eventBus.publish<TaskFolderCreatedEvent>({
      type: "TaskFolderCreated",
      taskId,
      folderPath,
      timestamp: Date.now(),
      // correlationId: command.correlationId,
    });

    // Emit TaskCreated event
    await this.eventBus.publish<TaskCreatedEvent>({
      type: "TaskCreated",
      taskId,
      taskName: command.taskName,
      config: command.taskConfig,
      timestamp: Date.now(),
      // correlationId: command.correlationId,
    });
    this.taskRepository.save(task);

    // Auto start task
    await this.eventBus.publish<StartTaskCommand>({
      type: "StartTaskCommand",
      taskId,
      timestamp: Date.now(),
      // correlationId: command.correlationId,
    });
  }

  private async handleStartTaskCommand(
    command: StartTaskCommand
  ): Promise<void> {
    const { taskId } = command;
    const task = this.taskRepository.findById(taskId);

    if (!task) {
      throw new Error(`Task ${taskId} not found`);
    }

    // Load task state first
    await this.eventBus.publish<TaskLoadedEvent>({
      type: "TaskLoaded",
      taskId,
      taskState: task,
      timestamp: Date.now(),
      // correlationId: command.correlationId,
    });

    // Get first subtask and start it
    const subtaskId = task.subtasks[0].id;
    const correlationId = crypto.randomUUID();
    await this.eventBus.publish<StartSubtaskCommand>({
      type: "StartSubtaskCommand",
      taskId,
      subtaskId,
      timestamp: Date.now(),
      correlationId,
    });
  }

  private async onNextSubtaskTriggered(
    event: NextSubtaskTriggeredEvent
  ): Promise<void> {
    const task = this.taskRepository.findById(event.taskId);
    if (!task) {
      throw new Error(`Task ${event.taskId} not found`);
    }

    const subtask = task.subtasks.find((s) => s.id === event.currentSubtaskId);
    if (!subtask) {
      throw new Error(`Subtask ${event.currentSubtaskId} not found`);
    }
    if (subtask.status !== "completed") {
      throw new Error(`Subtask ${event.currentSubtaskId} not completed yet`);
    }

    // Find next subtask if any
    const nextSubtask = task.subtasks.find(
      (s) => s.stepNumber === subtask.stepNumber + 1
    );
    if (nextSubtask) {
      await this.eventBus.publish<StartSubtaskCommand>({
        type: "StartSubtaskCommand",
        taskId: task.id,
        subtaskId: nextSubtask.id,
        timestamp: Date.now(),
        correlationId: crypto.randomUUID(),
      });
    } else {
      // All subtasks completed, mark task as completed
      task.status = "COMPLETED";
    }

    await this.saveTask(task);
  }

  private async onUserRenamedTaskFolder(
    event: TaskFolderRenamedEvent
  ): Promise<void> {
    /**
     * 1. Is the new name followed the task folder naming convention? If not, warn the user.
     * 2. Update the task folder path in task.
     */
    const { taskId } = event;
    const task = this.taskRepository.findById(taskId);
    if (!task) {
      throw new Error(`Task ${event.taskId} not found`);
    }

    // Update task with new folder path
    task.folderPath = event.newFolderPath;
    task.updatedAt = Date.now();

    // Save updated task
    await this.saveTask(task);
  }

  private async onUserDeletedTaskJson(): Promise<void> {
    /**
     * 1.
     */
  }

  private async saveTask(task: Task): Promise<void> {
    this.taskRepository.save(task);
    await this.workspaceFileManager.saveTaskToJson(task);
  }

  // private async onTaskCreated(event: TaskCreatedEvent): Promise<void> {
  //   const task: Task = {
  //     id: event.taskId,
  //     name: event.taskName,
  //     status: TaskStatus.CREATED,
  //     subtasks: [],
  //     config: event.config,
  //     createdAt: event.timestamp,
  //     updatedAt: event.timestamp,
  //   };

  //   this.taskRepository.save(task);
  // }

  // private async onTaskFolderCreated(
  //   event: TaskFolderCreatedEvent
  // ): Promise<void> {
  //   const task = this.tasks.get(event.taskId);
  //   if (task) {
  //     task.folderPath = event.folderPath;
  //     task.updatedAt = event.timestamp;
  //     this.tasks.set(event.taskId, task);
  //   }
  // }
}

File: examples-copy/playground/events/task_service.py
import logging
from datetime import datetime
from uuid import uuid4

from .event_bus import IEventBus
from .repositories import TaskRepository
from .type_defs import (
    CreateTaskCommand,
    EventType,
    NextSubtaskTriggeredEvent,
    Role,
    StartSubtaskCommand,
    StartTaskCommand,
    Subtask,
    SubtaskStatus,
    Task,
    TaskCreatedEvent,
    TaskFolderCreatedEvent,
    TaskLoadedEvent,
    TaskStatus,
)
from .workspace_file_manager import WorkspaceFileManager


class TaskService:
    """Service for managing task lifecycle and operations."""

    def __init__(
        self,
        event_bus: IEventBus,
        task_repository: TaskRepository,
        workspace_manager: WorkspaceFileManager,
    ):
        self._event_bus = event_bus
        self._task_repository = task_repository
        self._workspace_manager = workspace_manager
        self._logger = logging.getLogger(__name__)

        # Register command handlers
        self._event_bus.subscribe(
            EventType.CREATE_TASK_COMMAND, self._handle_create_task_command
        )
        self._event_bus.subscribe(
            EventType.START_TASK_COMMAND, self._handle_start_task_command
        )
        self._event_bus.subscribe(
            EventType.NEXT_SUBTASK_TRIGGERED_EVENT, self._on_next_subtask_triggered
        )

    async def _handle_create_task_command(self, command: CreateTaskCommand) -> None:
        """Handle the creation of a new task.

        This method:
        1. Creates a new task with initial subtasks
        2. Creates the task folder
        3. Saves task metadata
        4. Auto-starts the task
        """
        task_id = str(uuid4())
        current_time = datetime.now()

        # Create initial task object
        task = Task(
            id=task_id,
            seq_number=0,  # Will be set by workspace manager
            title=command.task_name,
            status=TaskStatus.CREATED,
            folder_path='',  # Will be set after folder creation
            subtasks=[
                # Initial planning subtask
                Subtask(
                    id=str(uuid4()),
                    task_id=task_id,
                    seq_number=0,
                    title='Planning',
                    description='Initial planning phase',
                    status=SubtaskStatus.PENDING,
                    team={'agent': Role.ASSISTANT, 'human': None},
                    input_type='string',
                    output_type='json',
                ),
                # Setup subtask
                Subtask(
                    id=str(uuid4()),
                    task_id=task_id,
                    seq_number=1,
                    title='Setup',
                    description='Setup initial configuration',
                    status=SubtaskStatus.PENDING,
                    team={'agent': Role.FUNCTION_EXECUTOR, 'human': Role.USER},
                    input_type='json',
                    output_type='json',
                ),
            ],
            config=command.task_config,
            created_at=current_time,
            updated_at=current_time,
        )

        # Create task folder
        folder_path = await self._workspace_manager.create_task_folder(task)
        task.folder_path = folder_path

        # Publish TaskFolderCreated event
        await self._event_bus.publish(
            TaskFolderCreatedEvent(
                task_id=task_id,
                folder_path=folder_path,
                timestamp=datetime.now(),
                correlation_id=command.correlation_id,
            )
        )

        # Save task and publish TaskCreated event
        self._task_repository.save(task)
        await self._event_bus.publish(
            TaskCreatedEvent(
                task_id=task_id,
                task_name=command.task_name,
                config=command.task_config,
                timestamp=datetime.now(),
                correlation_id=command.correlation_id,
            )
        )

        # Auto-start the task
        await self._event_bus.publish(
            StartTaskCommand(
                task_id=task_id,
                timestamp=datetime.now(),
                correlation_id=command.correlation_id,
            )
        )

    async def _handle_start_task_command(self, command: StartTaskCommand) -> None:
        """Handle starting a task.

        This method:
        1. Loads the task state
        2. Validates task exists
        3. Starts the first subtask
        """
        task = self._task_repository.find_by_id(command.task_id)
        if not task:
            raise ValueError(f'Task {command.task_id} not found')

        # Load task state
        await self._event_bus.publish(
            TaskLoadedEvent(
                task_id=command.task_id,
                task_state=task,
                timestamp=datetime.now(),
                correlation_id=command.correlation_id,
            )
        )

        # Start first subtask if available
        if task.subtasks:
            first_subtask = task.subtasks[0]
            await self._event_bus.publish(
                StartSubtaskCommand(
                    task_id=task.id,
                    subtask_id=first_subtask.id,
                    timestamp=datetime.now(),
                    correlation_id=command.correlation_id,
                )
            )

    async def _on_next_subtask_triggered(
        self, event: NextSubtaskTriggeredEvent
    ) -> None:
        """Handle transitioning to the next subtask.

        This method:
        1. Validates current subtask is completed
        2. Finds and starts next subtask if available
        3. Completes task if no more subtasks
        """
        task = self._task_repository.find_by_id(event.task_id)
        if not task:
            raise ValueError(f'Task {event.task_id} not found')

        current_subtask = next(
            (s for s in task.subtasks if s.id == event.current_subtask_id), None
        )
        if not current_subtask:
            raise ValueError(f'Subtask {event.current_subtask_id} not found')

        if current_subtask.status != SubtaskStatus.COMPLETED:
            raise ValueError(
                f'Current subtask {event.current_subtask_id} not completed'
            )

        # Find next subtask
        next_subtask = next(
            (
                s
                for s in task.subtasks
                if s.seq_number == current_subtask.seq_number + 1
            ),
            None,
        )

        if next_subtask:
            # Start next subtask
            await self._event_bus.publish(
                StartSubtaskCommand(
                    task_id=task.id,
                    subtask_id=next_subtask.id,
                    timestamp=datetime.now(),
                    correlation_id=event.correlation_id,
                )
            )
        else:
            # Complete task if no more subtasks
            task.status = TaskStatus.COMPLETED
            await self._save_task(task)

    async def _save_task(self, task: Task) -> None:
        """Save task to both repository and file system."""
        task.updated_at = datetime.now()
        self._task_repository.save(task)
        await self._workspace_manager.save_task_to_json(task)
